packages/js/uri-resolvers/src/cache/RequestSynchronizerResolver.ts

```diff
@@ -0,0 +1,87 @@
+import { UriResolver, UriResolverLike } from "../helpers";
+
+import {
+  IUriResolver,
+  Uri,
+  CoreClient,
+  IUriResolutionContext,
+  UriPackageOrWrapper,
+} from "@polywrap/core-js";
+import { Result } from "@polywrap/result";
+
+export class RequestSynchronizerResolver<TError>
+  implements IUriResolver<TError | Error> {
+  private requestCache: Map<
+    string,
+    Promise<Result<UriPackageOrWrapper, TError | Error>>
+  > = new Map();
+
+  constructor(private resolverToSynchronize: IUriResolver<TError>) {}
+
+  static from<TResolverError = unknown>(
+    resolver: UriResolverLike
+  ): RequestSynchronizerResolver<TResolverError> {
+    return new RequestSynchronizerResolver(
+      UriResolver.from<TResolverError>(resolver)
+    );
+  }
+
+  async tryResolveUri(
+    uri: Uri,
+    client: CoreClient,
+    resolutionContext: IUriResolutionContext
+  ): Promise<Result<UriPackageOrWrapper, TError | Error>> {
+    const existingRequest = this.requestCache.get(uri.uri);
+
+    if (existingRequest) {
+      return existingRequest.then(
+        (result) => {
+          resolutionContext.trackStep({
+            sourceUri: uri,
+            result,
+            description: "RequestSynchronizerResolver (Cache)",
+          });
+
+          return result;
+        },
+        () => {
+          // In case of an error, we try to resolve the URI again.
+          // This is because the error might be caused by a network issue or something similar,
+          // and we don't want all the requests to fail.
+          return this.tryResolveUri(uri, client, resolutionContext);
```

Comment:
Yep, this was an oversight.
Now, in case of a thrown exception the resolver will also throw (for all listeners)
In case of a resolution error, by default the resolver will ignore the cache for all but the first listener.
If the use wants to modify that behavior, they can use the `shouldUseCache` handler to use cache in case of a specific resolution error type (or all errors)